Sources/NIO/EventLoop.swift

```diff
@@ -50,6 +50,61 @@ public struct Scheduled<T> {
     }
 }
 
+public class RepeatedTask {
+    private let delay: TimeAmount
+    private var scheduled: Scheduled<EventLoopFuture<Void>>? {
+        didSet {
+            self.reschedule()
+        }
+    }
+    private var task: (() throws -> EventLoopFuture<Void>)?
+
+    internal init(interval: TimeAmount, task: @escaping () throws -> EventLoopFuture<Void>) {
+        self.delay = interval
+        self.task = task
+    }
+
+    /// Try to cancel the execution of the repeated task.
+    ///
+    /// Whether the execution of the task is immediately canceled depends on whether the execution of a task has already begun.
+    ///  This means immediate cancellation is not guaranteed.
+    public func cancel() {
+        self.scheduled?.cancel()
+        self.scheduled = nil
+        self.task = nil
+    }
+
+    internal func begin(on eventLoop: EventLoop, in delay: TimeAmount) {
+        guard let task = self.task else {
+            return
+        }
+        self.scheduled = eventLoop.scheduleTask(in: delay, task)
+    }
+
+    private func reschedule() {
+        self.scheduled?.futureResult.whenSuccess { future in
+            future.whenComplete {
+                self.reschedule0()
+            }
+        }
+        self.scheduled?.futureResult.whenFailure { error in
+            if case EventLoopError.cancelled = error {
+                return
+            }
+            if case EventLoopError.shutdown = error {
+                return
+            }
+            self.reschedule0()
```

Comment:
what other errors could we hit here?

Reply:
There could be an EventLoopError.unsupportedOperation, EventLoopError.shutdownFailed or any of the users thrown errors right? Should I handle the two EventLoopError cases?

Reply:
I don't think it can. If you `schedule` a function that returns a `Future`, the user thrown errors from that function cannot appear in the outer future.

Reply:
I think it may be fair to say that if the scheduled `futureResult` fails, there is no reason to reschedule, right?

Reply:
Yes with the current status the only way the `futureResult` fails is either during shutdown with `.shutdown` or if the task is cancelled with `cancelled`. No other error is possible at the moment. So I can remove this `whenFailure` case.

Reply:
if you want to be sure it's one of those errors you could just assert that it's one of them:

    assert(([EventLoopError.cancelled, EventLoopError.shutdown] as [Error]).contains(error), "unexpected error: \(error)")